### ISSUE TITLE
IPv6 Compliance RFC4861: Redirected Twice (Hosts Only) [v6LC.2.3.7]

### ISSUE THREAD

**Issue by LiveFreeAndRoam** [systemd/systemd#31981]
### systemd version the issue has been seen with

255 + commit 59b72d6f47160c74a1d7f44bc3e2373b4e14dd58 (HEAD -> network-next, origin/network-next)

### Used distribution

Ubuntu 23

### Linux kernel version used

6.5.0-17-generic

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd

### Expected behaviour you didn't see

> **NOTE**: This was initially recorded as a [comment](https://github.com/systemd/systemd/issues/31438#issuecomment-1975636499) in Issue https://github.com/systemd/systemd/issues/31438.  However, that was not giving this issue the correct visibility for tracking, as it stands alone and should be recorded separately.

#### v6LC.2.3.7: Redirected Twice (Hosts Only)
**Purpose**: Verify that a host properly processes valid Redirect messages twice for the same destination.
**Reference**: [RFC4861 - Sections 8.3](https://datatracker.ietf.org/doc/html/rfc4861#section-8.3)

#### Network Configuration
```
   TR1 TR2  TR3
    |   |   |
    +---+---+
    |       |
   TN1     HUT
```
- **TR** = Test Router
- **TN** = Test Node
- **HUT** = Host Under Test

#### Procedure
| Step | Action | Expected Behaviour of HUT |
|--------|--------|--------|
| 1 | TR1 forwards an Echo Request from TN1 to the HUT. The Source Address is the off-link global address of TN1. The Destination Address is the GA<sup>(1)</sup> of the HUT. | The HUT should respond to the Echo Request using TR1 as a first hop, as it is the only router in the HUT’s Default Router List. |
| 2 | TR1 sends `Redirect Dst="TN1 GA" Target="TR2 LLA"` to HUT. | |
| 3 | TR1 sends `Redirect Dst="TN1 GA" Target=TR2 LLA"` to HUT. | |
| 4 | TR1 forwards an Echo Request from TN1 to the HUT. The Source Address is the off-link global address of TN1. The Destination Address is the global address of the HUT. | The HUT should respond to the Echo Request using TR2 as a first hop, indicating the HUT processed the Redirect message. |
| 5 |  TR2 sends `Redirect Dst="TN1 GA" Target="TR3 LLA"` to HUT. | |
| 6 |  TR1 forwards an Echo Request from TN1 to the HUT. The Source Address is the off-link global address of TN1. The Destination Address is the global address of the HUT. | The HUT should respond to the Echo Request using TR3 as a first hop, indicating the HUT processed the Redirect message. |

* **(1)** - Global Address

### Unexpected behaviour you saw

In Step 6, the HUT responded to the Echo request using TR2, when it should have used TR3.

### Steps to reproduce the problem

[v6LC_2_3_07-Net0-20240304123454.zip](https://github.com/systemd/systemd/files/14476120/v6LC_2_3_07-Net0-20240304123454.zip)

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
#
# Preamble, setting everything up (ignore the preamble and jump to the next comment)
#
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 2h 30min
enp0s9: NDISC: Invoking callback for 'router' event.
Setting '/proc/sys/net/ipv6/neigh/enp0s9/base_reachable_time_ms' to '30000'
No change in value '30000', suppressing write
Setting '/proc/sys/net/ipv6/neigh/enp0s9/retrans_time_ms' to '1000'
No change in value '1000', suppressing write
Setting '/proc/sys/net/ipv6/conf/enp0s9/hop_limit' to '64'
No change in value '64', suppressing write
enp0s9: Requesting NDisc route (n/a): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc route (n/a): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc address (configured): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: Setting SLAAC addresses and router.
enp0s9: State changed: configured -> configuring
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: Received new foreign route (configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Configuring NDisc route (requesting): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Configuring NDisc route (requesting): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Configuring NDisc address (requesting,configured): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: Received new NDisc route (configured): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Received updated NDisc address (configured): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: ndisc_check_ready(): SLAAC addresses and routes are not set.
enp0s9: ndisc_check_ready(): SLAAC addresses and routes are not set.
enp0s9: Received updated NDisc address (configured): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: SLAAC addresses and routes set.
enp0s9: link_check_ready(): IPv4LL:no DHCPv4:no DHCPv6:no DHCP-PD:no NDisc:yes
enp0s9: State changed: configuring -> configured
enp0s9: NDISC: Received Neighbor Advertisement from fe80::200:10ff:fe10:1060: Router=yes, Solicited=yes, Override=no
enp0s9: NDISC: Invoking callback for 'neighbor' event.
enp0s9: NDISC: Received Neighbor Advertisement from fe80::200:10ff:fe10:1061: Router=yes, Solicited=yes, Override=no
enp0s9: NDISC: Invoking callback for 'neighbor' event.
enp0s9: NDISC: Received Neighbor Advertisement from fe80::200:10ff:fe10:1062: Router=yes, Solicited=yes, Override=no
enp0s9: NDISC: Invoking callback for 'neighbor' event.
#
# First Redirect results in new route.
#
enp0s9: NDISC: Received Redirect message from fe80::200:10ff:fe10:1060: Target=fe80::200:10ff:fe10:1061, Destination=2001:2:0:1001:200:10ff:fe10:1180
enp0s9: NDISC: Invoking callback for 'redirect' event.
enp0s9: Requesting NDisc route (n/a): dst: 2001:2:0:1001:200:10ff:fe10:1180/128, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: redirect, scope: global, type: unicast, flags: n/a
enp0s9: Configuring NDisc route (requesting): dst: 2001:2:0:1001:200:10ff:fe10:1180/128, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: redirect, scope: global, type: unicast, flags: n/a
enp0s9: Received new NDisc route (configured): dst: 2001:2:0:1001:200:10ff:fe10:1180/128, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: redirect, scope: global, type: unicast, flags: n/a
enp0s9: SLAAC addresses and routes set.

#
# This is the 2nd redirect. No route is configured for this 2nd Redirect.
#
enp0s9: NDISC: Received Redirect message from fe80::200:10ff:fe10:1061: Target=fe80::200:10ff:fe10:1062, Destination=2001:2:0:1001:200:10ff:fe10:1180
enp0s9: NDISC: Invoking callback for 'redirect' event.
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 0
enp0s9: NDISC: Invoking callback for 'router' event.
enp0s9: Removing NDisc route (configured): dst: 2001:2:0:1001:200:10ff:fe10:1180/128, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: redirect, scope: global, type: unicast, flags: n/a
enp0s9: Removing foreign route (configuring,configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Found an existing route that conflicts with new route based on a received RA, removing.
enp0s9: Removing NDisc route (configured): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc route (configured,removing): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Setting SLAAC addresses and router.
enp0s9: Forgetting removed foreign route (configuring): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Refuse to configure NDisc route with zero lifetime.
enp0s9: NDISC: Received Neighbor Advertisement from fe80::200:10ff:fe10:1060: Router=yes, Solicited=no, Override=yes
enp0s9: NDISC: Invoking callback for 'neighbor' event.
enp0s9: Forgetting removed NDisc route (n/a): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 0
enp0s9: NDISC: Invoking callback for 'router' event.
enp0s9: Requesting NDisc route (n/a): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Setting SLAAC addresses and router.
enp0s9: State changed: configured -> configuring
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: Forgetting removed NDisc route (n/a): dst: 2001:2:0:1001:200:10ff:fe10:1180/128, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: redirect, scope: global, type: unicast, flags: n/a
enp0s9: Refuse to configure NDisc route with zero lifetime.
enp0s9: NDISC: Received Neighbor Advertisement from fe80::200:10ff:fe10:1061: Router=yes, Solicited=no, Override=yes
enp0s9: NDISC: Invoking callback for 'neighbor' event.
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 0
enp0s9: NDISC: Invoking callback for 'router' event.
enp0s9: Requesting NDisc route (n/a): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Setting SLAAC addresses and router.
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
Could not drop route, ignoring: No such process
enp0s9: Refuse to configure NDisc route with zero lifetime.
enp0s9: NDISC: Received Neighbor Advertisement from fe80::200:10ff:fe10:1062: Router=yes, Solicited=no, Override=yes
enp0s9: NDISC: Invoking callback for 'neighbor' event.
Could not drop route, ignoring: No such process
```

**Comment by yuwata** [comment]
I guess you tested in something wrong way. The test case sends redirect message three times, but you seem to check only up to the second message. The second message is equivalent to the first one. So adding no new route on second is expected.

**Comment by LiveFreeAndRoam** [comment]
Thanks. I'll rerun the test on Tue. We have Easter holidays here, now.

**Comment by yuwata** [comment]
What's the current status of this? Could you re-test with the current main branch?

**Comment by LiveFreeAndRoam** [comment]
It still fails.  I tested with:

```
$ git log -n 1
commit 11a150bc43d08e4d1e1622aca696770c59bda03c (HEAD -> main, origin/main, origin/HEAD)
```

**Comment by yuwata** [comment]
Ah, I misunderstood the test.
Currently, Redirect message is only processed if the sender is the main router. I will re-read RFC and rework on the issue.

**Comment by yuwata** [comment]
@LiveFreeAndRoam 
If I read [RFC 4861 section 8.1](https://datatracker.ietf.org/doc/html/rfc4861#section-8.1) correctly, the Redirect message in step 5 should be dropped:
> The IP source address of the Redirect is the same as the current first-hop router for the specified ICMP Destination Address.

Is the statement has been updated/replaced/drooped in a later RFC?? Or do I misunderstand the statement??

**Comment by yuwata** [comment]
Ah, I understand the statement. the first-hop router does not mean the default router, but per-destination first-hop router.

**Comment by yuwata** [comment]
Fix is waiting in #32213. The same commits are also in the network-next branch.

**Comment by LiveFreeAndRoam** [comment]
This test passes! Nice one, @yuwata.

Tested with:

```
$ git log -n 1
commit 9a0602be3401770af629817f7e834fd7118bab46 (HEAD -> network-next, origin/network-next)
```

[v6LC_2_3_07-Net0-20240304132449.zip](https://github.com/systemd/systemd/files/14942158/v6LC_2_3_07-Net0-20240304132449.zip)
